native_store_test.go

```diff
@@ -185,3 +209,133 @@ func TestNewDefaultNativeStore(t *testing.T) {
 		t.Errorf("NewDefaultNativeStore() = %v, want %v", ok, wantOK)
 	}
 }
+
+func TestNativeStore_trace(t *testing.T) {
+	ns := &nativeStore{
+		&testExecuter{},
+	}
+	// create trace hooks that write to buffer
+	buffer := bytes.Buffer{}
+	traceHook := &trace.ExecutableTrace{
+		ExecuteStart: func(executableName string, action string) {
+			buffer.WriteString(fmt.Sprintf("test trace, start the execution of executable %s with action %s ", executableName, action))
+		},
+		ExecuteDone: func(executableName string, action string, err error) {
+			buffer.WriteString(fmt.Sprintf("test trace, completed the execution of executable %s with action %s and got err %v", executableName, action, err))
+		},
+	}
+	ctx := trace.WithExecutableTrace(context.Background(), traceHook)
+	// Test ns.Put trace
+	err := ns.Put(ctx, traceHost, auth.Credential{Username: testUsername, Password: testPassword})
+	if err != nil {
+		t.Fatalf("trace test ns.Put fails: %v", err)
+	}
+	bufferContent := buffer.String()
+	if bufferContent != "test trace, start the execution of executable testExecuter with action store test trace, completed the execution of executable testExecuter with action store and got err <nil>" {
+		t.Fatalf("incorrect buffer content: %s", bufferContent)
+	}
+	buffer.Reset()
+	// Test ns.Get trace
+	_, err = ns.Get(ctx, traceHost)
+	if err != nil {
+		t.Fatalf("trace test ns.Get fails: %v", err)
+	}
+	bufferContent = buffer.String()
+	if bufferContent != "test trace, start the execution of executable testExecuter with action get test trace, completed the execution of executable testExecuter with action get and got err <nil>" {
+		t.Fatalf("incorrect buffer content: %s", bufferContent)
+	}
+	buffer.Reset()
+	// Test ns.Delete trace
+	err = ns.Delete(ctx, traceHost)
+	if err != nil {
+		t.Fatalf("trace test ns.Delete fails: %v", err)
+	}
+	bufferContent = buffer.String()
+	if bufferContent != "test trace, start the execution of executable testExecuter with action erase test trace, completed the execution of executable testExecuter with action erase and got err <nil>" {
+		t.Fatalf("incorrect buffer content: %s", bufferContent)
+	}
+}
+
+// This test ensures that a nil trace will not cause an error.
+func TestNativeStore_noTrace(t *testing.T) {
```

Comment:
This test case looks identical with `TestNativeStore_basicAuth`, which tests the scenario where the context carries no trace. You may also want to test the case where the context carries an empty trace.

Reply:
Added the empty trace test case. Good suggestion, thanks.